File: backend-nestjs/src/services/public-media-url.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ApplicationEntity } from '../entities/application.entity';

@Injectable()
export class PublicMediaUrlService {
  private readonly contentPlatformBaseUrl: string;
  private readonly minioPublicBaseUrl: string;

  constructor(private readonly config: ConfigService) {
    this.contentPlatformBaseUrl = this.normalizeBaseUrl(
      this.config.get<string>('CONTENT_PLATFORM_BASE_URL') ||
      this.config.get<string>('PUBLIC_BASE_URL') ||
      'http://localhost:3000',
    );
    this.minioPublicBaseUrl = this.normalizeBaseUrl(
      this.config.get<string>('MINIO_PUBLIC_BASE_URL') || 'http://localhost:9000',
    );
  }

  getPublicBaseUrl(application: ApplicationEntity): string {
    return this.normalizeBaseUrl(application.publicBaseUrlOverride || this.contentPlatformBaseUrl);
  }

  toPublicMediaUrl(application: ApplicationEntity, inputUrlOrPath: string | null): string | null {
    const value = inputUrlOrPath?.trim();
    if (!value) {
      return null;
    }

    const publicBaseUrl = this.getPublicBaseUrl(application);
    const mediaPath = this.extractMediaPath(value);
    if (mediaPath) {
      return `${publicBaseUrl}${mediaPath}`;
    }

    const objectKeyPath = this.extractApplicationObjectKeyPath(application, value);
    if (objectKeyPath) {
      return `${publicBaseUrl}${objectKeyPath}`;
    }

    if (this.isAbsoluteUrl(value)) {
      if (this.startsWithBaseUrl(value, this.minioPublicBaseUrl)) {
        return null;
      }
      return value.replace(/\/$/, '');
    }

    if (value.startsWith('/')) {
      return `${publicBaseUrl}${value}`;
    }

    return `${publicBaseUrl}/${value}`;
  }

  rewriteHtmlMediaUrls(application: ApplicationEntity, html: string | null): string | null {
    if (!html) {
      return html;
    }

    return html.replace(/\b(src|href)\s*=\s*(["'])([^"']+)\2/gi, (_match, attribute: string, quote: string, url: string) => {
      const mediaPath = this.extractMediaPath(url);
      const objectKeyPath = this.extractApplicationObjectKeyPath(application, url);
      if (!mediaPath && !objectKeyPath) {
        return `${attribute}=${quote}${url}${quote}`;
      }
      return `${attribute}=${quote}${this.getPublicBaseUrl(application)}${mediaPath || objectKeyPath}${quote}`;
    });
  }

  private extractMediaPath(value: string): string | null {
    const normalized = value.trim();
    if (!normalized) {
      return null;
    }

    if (normalized.startsWith('/media/')) {
      return normalized;
    }
    if (normalized.startsWith('media/')) {
      return `/${normalized}`;
    }

    const mediaIndex = normalized.indexOf('/media/');
    if (mediaIndex >= 0) {
      return normalized.slice(mediaIndex);
    }

    if (this.startsWithBaseUrl(normalized, this.minioPublicBaseUrl)) {
      return this.extractFromBase(normalized, this.minioPublicBaseUrl);
    }
    if (this.startsWithBaseUrl(normalized, this.contentPlatformBaseUrl)) {
      return this.extractFromBase(normalized, this.contentPlatformBaseUrl);
    }

    if (this.isAbsoluteUrl(normalized)) {
      try {
        const parsed = new URL(normalized);
        if (parsed.pathname.startsWith('/media/')) {
          return `${parsed.pathname}${parsed.search}${parsed.hash}`;
        }
      } catch {
        return null;
      }
    }

    return null;
  }

  private extractApplicationObjectKeyPath(application: ApplicationEntity, value: string): string | null {
    const normalized = value.trim();
    if (!normalized) {
      return null;
    }

    const objectKey = this.extractObjectKey(application.id, normalized);
    return objectKey ? `/media/${objectKey}` : null;
  }

  private extractObjectKey(applicationId: string, value: string): string | null {
    if (value.startsWith(`${applicationId}/`)) {
      return value;
    }

    if (this.startsWithBaseUrl(value, this.minioPublicBaseUrl)) {
      try {
        const parsed = new URL(value);
        const path = parsed.pathname.replace(/^\/+/, '');
        const withoutBucket = path.startsWith('media/') ? path.slice('media/'.length) : path;
        return withoutBucket.startsWith(`${applicationId}/`) ? withoutBucket : null;
      } catch {
        return null;
      }
    }

    return null;
  }

  private extractFromBase(value: string, baseUrl: string): string | null {
    const withoutBase = value.slice(baseUrl.length);
    if (!withoutBase.startsWith('/media/')) {
      return null;
    }
    return withoutBase;
  }

  private normalizeBaseUrl(url: string): string {
    return url.trim().replace(/\/+$/, '');
  }

  private isAbsoluteUrl(value: string): boolean {
    return /^https?:\/\//i.test(value);
  }

  private startsWithBaseUrl(value: string, baseUrl: string): boolean {
    return value.toLowerCase().startsWith(baseUrl.toLowerCase());
  }
}

File: backend/src/main/java/com/contentplatform/backend/application/service/PublicMediaUrlService.java
package com.contentplatform.backend.application.service;

import com.contentplatform.backend.application.port.out.ApplicationRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class PublicMediaUrlService {
    private static final Pattern HTML_LINK_PATTERN = Pattern.compile("(?i)(src|href)=(['\"])([^'\"]+)\\2");

    private final ApplicationRepository applicationRepository;
    private final String contentPlatformBaseUrl;
    private final String minioPublicBaseUrl;

    public PublicMediaUrlService(ApplicationRepository applicationRepository,
                                 @Value("${app.delivery.content-platform-base-url:http://localhost:3000}") String contentPlatformBaseUrl,
                                 @Value("${app.delivery.minio-public-base-url:http://localhost:9000}") String minioPublicBaseUrl) {
        this.applicationRepository = applicationRepository;
        this.contentPlatformBaseUrl = normalizeBase(contentPlatformBaseUrl, "http://localhost:3000");
        this.minioPublicBaseUrl = normalizeBase(minioPublicBaseUrl, "http://localhost:9000");
    }

    public String getPublicBaseUrl(String applicationId) {
        return applicationRepository.findById(applicationId)
            .map(application -> application.getWebsiteUrl())
            .filter(url -> url != null && !url.isBlank())
            .map(this::normalizeBase)
            .orElse(contentPlatformBaseUrl);
    }

    public String toPublicMediaUrl(String applicationId, String inputUrlOrPath) {
        String mediaPath = extractMediaPath(inputUrlOrPath);
        if (mediaPath == null) {
            return null;
        }
        return getPublicBaseUrl(applicationId) + mediaPath;
    }

    public String rewriteHtmlMediaUrls(String applicationId, String html) {
        if (html == null || html.isBlank()) {
            return html;
        }
        Matcher matcher = HTML_LINK_PATTERN.matcher(html);
        StringBuffer rewritten = new StringBuffer();
        while (matcher.find()) {
            String replacementUrl = toPublicMediaUrl(applicationId, matcher.group(3));
            if (replacementUrl == null) {
                matcher.appendReplacement(rewritten, Matcher.quoteReplacement(matcher.group(0)));
                continue;
            }
            String replacement = matcher.group(1) + "=" + matcher.group(2) + replacementUrl + matcher.group(2);
            matcher.appendReplacement(rewritten, Matcher.quoteReplacement(replacement));
        }
        matcher.appendTail(rewritten);
        return rewritten.toString();
    }

    private String extractMediaPath(String inputUrlOrPath) {
        if (inputUrlOrPath == null) {
            return null;
        }
        String trimmed = inputUrlOrPath.trim();
        if (trimmed.isEmpty()) {
            return null;
        }
        if (trimmed.startsWith("/media/")) {
            return trimmed;
        }
        if (trimmed.startsWith("media/")) {
            return "/" + trimmed;
        }
        if (trimmed.startsWith(minioPublicBaseUrl + "/media/")) {
            return trimmed.substring(minioPublicBaseUrl.length());
        }
        if (trimmed.startsWith(contentPlatformBaseUrl + "/media/")) {
            return trimmed.substring(contentPlatformBaseUrl.length());
        }
        try {
            URI uri = URI.create(trimmed);
            String path = uri.getPath();
            if (path != null) {
                int mediaIndex = path.indexOf("/media/");
                if (mediaIndex >= 0) {
                    return appendQueryAndFragment(path.substring(mediaIndex), uri.getQuery(), uri.getFragment());
                }
            }
        } catch (IllegalArgumentException ignored) {
            // Fall back to raw string matching for legacy malformed values.
        }
        int mediaIndex = trimmed.indexOf("/media/");
        if (mediaIndex >= 0) {
            return trimmed.substring(mediaIndex);
        }
        return null;
    }

    private String appendQueryAndFragment(String path, String query, String fragment) {
        StringBuilder builder = new StringBuilder(path);
        if (query != null && !query.isBlank()) {
            builder.append('?').append(query);
        }
        if (fragment != null && !fragment.isBlank()) {
            builder.append('#').append(fragment);
        }
        return builder.toString();
    }

    private String normalizeBase(String baseUrl) {
        return normalizeBase(baseUrl, "http://localhost:3000");
    }

    private String normalizeBase(String baseUrl, String fallback) {
        String trimmed = (baseUrl == null || baseUrl.isBlank()) ? fallback : baseUrl.trim();
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        return trimmed;
    }
}
